core/ordersync/ordersync.go

```diff
@@ -0,0 +1,507 @@
+	// found anywhere on the network. This can mean that we aren't connected to any
+	// peers on the same topic, that there are no orders for the topic throughout
+	// the entire network, or that there are peers that have the orders we're
+	// looking for, but they are refusing to give them to us.
+	ErrNoOrders = errors.New("no orders where received from any known peers")
+)
+
+// NoMatchingSubprotocolsError is returned whenever two peers attempting to use
+// the ordersync protocol cannot agree on a subprotocol to use.
+type NoMatchingSubprotocolsError struct {
+	Requested []string
+	Supported []string
+}
+
+func (e NoMatchingSubprotocolsError) Error() string {
+	return fmt.Sprintf("could not agree on an ordersync subprotocol (requested: %v, supported: %s)", e.Requested, e.Supported)
+}
+
+const (
+	// ID is the ID for the ordersync protocol.
+	ID = protocol.ID("/0x-mesh/order-sync/version/0")
+)
+
+// Request represents a high-level ordersync request. It abstracts away some
+// of the details of subprotocol negotiation and encoding/decoding.
+type Request struct {
+	RequesterID peer.ID     `json:"requesterID"`
+	Metadata    interface{} `json:"metadata"`
+}
+
+// rawRequest contains all the details we need at the lowest level to encode/decode
+// the request and perform subprotocol negoatiation.
+type rawRequest struct {
+	Type         string          `json:"type"`
+	Subprotocols []string        `json:"subprotocols"`
+	Metadata     json.RawMessage `json:"metadata"`
+}
+
+// Response represents a high-level ordersync response. It abstracts away some
+// of the details of subprotocol negotiation and encoding/decoding.
+type Response struct {
+	ProviderID peer.ID               `json:"providerID"`
+	Orders     []*zeroex.SignedOrder `json:"orders"`
+	Complete   bool                  `json:"complete"`
+	Metadata   interface{}           `json:"metadata"`
+}
+
+// rawResponse contains all the details we need at the lowest level to encode/decode
+// the response, perform subprotocol negoatiation, and more.
+type rawResponse struct {
+	Type        string                `json:"type"`
+	Subprotocol string                `json:"subprotocol"`
+	Orders      []*zeroex.SignedOrder `json:"orders"`
+	Complete    bool                  `json:"complete"`
+	Metadata    json.RawMessage       `json:"metadata"`
+}
+
+// Service is the main entrypoint for running the ordersync protocol. It handles
+// responding to and sending ordersync requests.
+type Service struct {
+	ctx          context.Context
+	node         *p2p.Node
+	subprotocols map[string]Subprotocol
+	// requestRateLimiter is a rate limiter for incoming ordersync requests. It's
+	// shared between all peers.
+	requestRateLimiter *rate.Limiter
+}
+
+// SupportedSubprotocols returns the subprotocols that are supported by the service.
+func (s *Service) SupportedSubprotocols() []string {
+	sids := []string{}
+	for sid := range s.subprotocols {
```

Comment:
This snippet suggests that this array can have a random order (unless I'm missing something): https://play.golang.org/p/s3WTo2LVuyl. With this in mind, we aren't actually enforcing a preference on these subprotocols. What do you think about keeping the original subprotocols array in the `Service` object?

Reply:
Arrays and slices do not have randomized order. Maps have randomized order. When we implement more subprotocols we just need to add an additional field to `Service` to keep track of the original order for the subprotocols that were passed in.

Reply:
Your last sentence was the point of my comment. Is there a reason not to fix this now though?